text/0053-enum-entities.md

```diff
@@ -0,0 +1,146 @@
+```
+permit(
+    principal,
+    action == Action::"UpdateTask",
+    resource)
+when {
+    resource.status != Color.red
+};
+```
+This syntax is similar to what's provided for Java `enum`s. 
+
+The benefit of this approach is that it may feel a little more natural than representing a concept, like a color, as a set of legal entity values. It would also be easy to encode this approach in a policy analysis.
+
+The main drawback of this approach is that it introduces a new primitive type to the language that does not make much sense without schemas. Recall that for Cedar, validation with a schema is _optional_. We'd allow users to include any random enumerated identifiers (like Scheme-style _symbols_) in policies which, without validation, would fail equality checks at run-time.
+
+The proposed approach that lists particular entity values as an enumeration makes no change to the language, leveraging the existing entity type concept. So policies without schemas still make sense. With the added schema, there is additional benefit when validating and constructing reasonable policies.
+
+### Alternative C: Enumerated entities with hierarchy
+
+Enumerated entities as proposed are limited in their functionality and specificity. We could extend them. For example:
+```
+entity Application enum [ "TinyTodo" ];
+entity RequestEntity enum [ "principal", "resource" ] in [ Application ];
+entity User in [ Application ];
+action CreateList
+    appliesTo { principal: [User], resource: [Application] };
+action GetLists
+    appliesTo { principal: [User, RequestEntity::"principal"],
+                resource: [Application]};
+```
+This differs from some earlier examples in the following ways:
+1. Enumerated entities can have parents in the entity hierarchy, and can be parents of other enumerated entity values; both cases are shown in the definition of `RequestEntity`
+2. Enumerated entities can appear as _singleton types_, e.g., as `RequestEntity::"principal"` in the definition of action `GetLists`.
+
+Both of these extensions are similar to what's available for `Action`s right now, but generalized to arbitrary entity types. You could also imagine enumerated entity types having attributes, as has been anticipated for `Action`s.
+
+One drawback of this Alternative is that it creates added complication for both the validator (e.g., to handle singleton types) and the analyzer (e.g., to deal with the more general hierarchy constraints).
+
+Another drawback is that it adds complication to the entity store: Once you add hierarchy constraints and attributes, you need to create actual entities to include with policy requests (or extract them from the schema at request-time). The RFC as proposed does not require that.
+
+The good news is that this Alternative is a strict generalization of the RFC as proposed, which means if we agree to the current proposal we can later upgrade to some or all of this Alternative without incurring a breaking change.
```

Comment:
I wouldn't be surprised if we re-thought Actions in schema while implementing this extension, so we may end up _wanting_ to make a breaking change, but that could come up regardless of when we implement this part. 

Reply:
Adding ancestors and attributes is okay for analysis as long as they are all fully known and specified in the schema.  What would be difficult to handle is if we wanted to have some `enum` instance (or, for that matter, some action) with unknown or partially known ancestors or attribute.   

I don’t see a good use case for partially specified actions and enums but it would be good to clarify in the RFC that everything about these special entities (that are fixed in the schema) must be fully specified.

Reply:
To be clear, this is okay in the sense that it doesn’t affect soundness, completeness, or decidability.  But it will have a negative effect on performance for large enums (see [this comment](https://github.com/cedar-policy/rfcs/pull/53#issuecomment-1974227827)).